crate/findex/src/batcher_findex.rs

```diff
@@ -0,0 +1,381 @@
+use crate::adt::BatchedIndexADT;
+use crate::error::BatchFindexError;
+use crate::memory_layers::batching_layer::{BatcherArc, MemoryBatcher};
+use crate::{Decoder, Encoder, Findex, IndexADT};
+use cosmian_sse_memories::{ADDRESS_LENGTH, Address, BatchingMemoryADT};
+use std::sync::atomic::AtomicUsize;
+use std::{collections::HashSet, fmt::Debug, hash::Hash, sync::Arc};
+
+#[derive(Debug)]
+pub struct BatcherFindex<
+    const WORD_LENGTH: usize,
+    Value: Send + Hash + Eq,
+    EncodingError: Send + Debug,
+    BatcherMemory: Clone + Send + BatchingMemoryADT<Address = Address<ADDRESS_LENGTH>, Word = [u8; WORD_LENGTH]>,
+> {
+    memory: BatcherMemory,
+    encode: Arc<Encoder<Value, BatcherMemory::Word, EncodingError>>,
+    decode: Arc<Decoder<Value, BatcherMemory::Word, EncodingError>>,
+}
+
+impl<
+    const WORD_LENGTH: usize,
+    Value: Send + Hash + Eq,
+    BatcherMemory: Debug
+        + Send
+        + Sync
+        + Clone
+        + BatchingMemoryADT<Address = Address<ADDRESS_LENGTH>, Word = [u8; WORD_LENGTH]>,
+    EncodingError: Send + Debug + std::error::Error,
+> BatcherFindex<WORD_LENGTH, Value, EncodingError, BatcherMemory>
+{
+    pub fn new(
+        memory: BatcherMemory,
+        encode: Encoder<Value, BatcherMemory::Word, EncodingError>,
+        decode: Decoder<Value, BatcherMemory::Word, EncodingError>,
+    ) -> Self {
+        Self {
+            memory,
+            encode: Arc::new(encode),
+            decode: Arc::new(decode),
+        }
+    }
+
+    // Insert or delete are both an unbounded number of calls to `guarded_write` on the memory layer.
+    async fn batch_insert_or_delete<Keyword>(
+        &self,
+        entries: Vec<(Keyword, impl Send + IntoIterator<Item = Value>)>,
+        is_insert: bool,
+    ) -> Result<(), BatchFindexError<BatcherMemory>>
+    where
+        Keyword: Send + Sync + Hash + Eq,
+    {
+        let mut search_futures = Vec::new();
+        let n = entries.len();
+        let buffered_memory = BatcherArc::new(MemoryBatcher::new_writer(
+            self.memory.clone(),
+            AtomicUsize::new(n),
+        ));
+
+        for (guard_keyword, bindings) in entries {
+            let memory_arc = buffered_memory.clone();
```

Comment:
This is bad practice to encode types in tames. Use shadowing instead. Plus what is a buffered memory? Isn't it an implementation detail?
```suggestion
            let batch_memory = batch_memory.clone();
```

Reply:
fixed